torch/csrc/jit/codegen/cuda/kernel_db/kernel_db.cpp

```diff
@@ -0,0 +1,253 @@
+#include <mutex>
+#include <regex>
+
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include "torch/csrc/jit/codegen/cuda/kernel_db/kernel_db.h"
+#include "torch/csrc/jit/codegen/cuda/kernel_db/utils.h"
+#include "torch/csrc/jit/codegen/cuda/utils.h"
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+static std::mutex kernel_db_lock;
+
+KernelDb::KernelDb(bool _disabled)
+    : disabled_(_disabled),
+      initialized_(false),
+      kernel_map_(),
+      kernel_db_path_(),
+      kernel_db_txt_file_() {}
+
+bool KernelDb::open(
+    const std::string& kernel_db_dir,
+    const std::string& kernel_db_file,
+    bool use_temp_dir) {
+  FUSER_PERF_SCOPE("KernelDb::open");
+  const std::string header(
+      "kernel_name,compile_args,kernel_code_file,cubin_file");
+
+  // The KernelDb directory is queried and created if it doesn't exist
+  {
+    FUSER_PERF_SCOPE("KernelDb::open::create_directory");
+    if (use_temp_dir) {
+      kernel_db_path_ = fs::temp_directory_path() / kernel_db_dir;
+    } else {
+      kernel_db_path_ = fs::path(kernel_db_dir);
+    }
+    if (!fs::is_directory(kernel_db_path_)) {
+      try {
+        fs::create_directory(kernel_db_path_);
+      } catch (const std::exception& e) {
+        TORCH_WARN(
+            "Unable to create nvFuser Kernel DB directory! ",
+            kernel_db_path_.string(),
+            e.what());
+        return false;
+      }
+    }
+  }
+
+  // The CSV file that captures the db is read if it exists
```

Comment:
This is great for now, but sounds a bit expensive, may want to offload this process on a thread and ignore the kernelDB until it's loaded. In the future we might want to use a database that doesn't need to read the entire file into memory.